modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java

```diff
@@ -739,33 +738,47 @@ public CheckpointHistoryResult searchAndReserveCheckpoints(
 
         CheckpointEntry oldestCpForReservation = null;
 
-        synchronized (earliestCp) {
+        synchronized (earliestCpGrps) {
             CheckpointEntry oldestHistCpEntry = firstCheckpoint();
 
-            for (Integer grpId : groupsAndPartitions.keySet()) {
+            for (Map.Entry<Integer, Set<Integer>> e0 : groupsAndPartitions.entrySet()) {
                 CheckpointEntry oldestGrpCpEntry = null;
 
-                for (Integer part : groupsAndPartitions.get(grpId)) {
-                    CheckpointEntry cpEntry = earliestCp.get(new GroupPartitionId(grpId, part));
+                Integer grpId = e0.getKey();
+
+                CacheGroupContext grpCtx = cacheGrpCtxSupplier.get(grpId);
+
+                if (grpCtx == null || !earliestCpGrps.contains(grpId))
+                    continue;
+
+                for (Integer partId : e0.getValue()) {
+                    GridDhtLocalPartition localPart = grpCtx.topology().localPartition(partId);
+
+                    if (localPart == null)
+                        continue;
+
+                    CheckpointEntry cp = histMap.get(localPart.earliestCpTs());
```

Comment:
Why do you not skip 0 here? I think, it is possible to have no history for a specific partition.

Reply:
Added check for skipping.